docs/raw/push-notification-server.md

```diff
@@ -0,0 +1,672 @@
+background service.
+Therefore Status provides a set of Push notification services that can be used
+to achieve this functionality.
+
+Because this can't be safely implemented in a privacy preserving manner, clients
+MUST be given an option not to enable push notifications and not to send push notifications.
+
+## Requirements
+
+Status needs to run their own [gorush](https://github.com/appleboy/gorush) 
+publicly accessible server for sending the actual notification. This cannot be community run as it requires sharing private credentials that only status has access to.
+
+## Registering with the push notification service
+
+A client MAY register with a Push Notification service of their choice.
+
+[//]: # (We can have a separate partitioned topic scheme, or individual based on PK)
+
+A message MUST be sent to the [partitioned topic](#partitioned-topic) for the
+public key of the node, encrypted with this key.
+
+The content of the message MUST contain the following [protobuf record](https://developers.google.com/protocol-buffers/):
+
+```protobuf
+
+message PushNotificationOptions {
+  boolean enabled = 1;
+  repeated string allowed_user_list = 2;
+  repeated string block_user_list = 3;
+  repeated string block_chat_list = 4;
+}
+
+message PushNotificationDeviceToken {
+  enum TokenType {
+    UNKWOWN_TOKEN_TYPE = 0;
+    APS_TOKEN = 1;
+    FIREBASE_TOKEN = 2;
+  }
+  TokenType token_type = 1;
+  string token = 2;
+  string installation_id = 3;
+  PushNotificationPreferences preferences = 4;
+}
+
+message PushNotificationPreferences {
+  repeated DeviceToken device_tokens = 1;
+  uint version = 2;
+  boolean unregister = 3;
+}
+
+message PushNotificationRegister {
+  bytes payload = 1;
+  bytes signature = 2;
+}
+```
+
+A push notification server will handle the message according to the following rules:
+
+- it MUST verify that the `signature` matches the public key of the sender.
+- it MUST verify that `token_type` is supported
+- it MUST verify that `token` is non empty
+- it MUST verify that `installation_id` is non empty
+- it MUST verify that `version` is non-zero and greater than the currently stored, if any
+- it MUST verify that `device_tokens` is non empty
+
+If `signature` does not match the public key of the sender, the message MUST be discarded.
+
+If `token_type` is not supported, a response MUST be sent with `error` set to 
+`UNSUPPORTED_TOKEN_TYPE`.
+
+If `token`,`installation_id`,`device_tokens`,`version` are empty, a response MUST
+be sent with `error` set to `MALFORMED_MESSAGE`.
+
+If the `version` is equal or less than the currently stored version, a response MUST
+be sent with `error` set to `VERSION_MISMATCH`.
+
+If any other error occurs the `error` should be set to `INTERNAL_ERROR`.
+
+If the request is successful, an access token MUST be generated and sent back to the client
+with `success` set to `true`.
+
+It is RECOMMENDED to use a randomly generated string of at least 36 characters.
```

Comment:
for which field?

Reply:
I lost a sentence somewhere :) it's for the access token